## <font color=darkblue> European Vanilla Option pricing using Binomial Tree</font>

### Pricing a European Call option: 
* CMP = 95
* K = 105
* T = 1 year
* r = 0

We will use a three steps binomial tree

In [1]:
call, put, european, american = 100, 101, 102, 103

In [2]:
side = call  # Option side
style = european  # Option style
price = 95  #Current instrument price
strike = 105  #Strike price
riskfree = 0  
divyield = 0
tte = 365  # Time to expiration in days

print('Calculation Inputs')
print('%18s : %0.3f' % ('Price', price))
print('%18s : %0.3f' % ('Strike', strike))
print('%18s : %0.3f' % ('Risk-free', riskfree))
print('%18s : %0.3f' % ('Div Yield', divyield))
print('%18s : %0.3f' % ('TTE Days', tte))
print()

Calculation Inputs
             Price : 95.000
            Strike : 105.000
         Risk-free : 0.000
         Div Yield : 0.000
          TTE Days : 365.000



Setting the upper movement of the price <br>
u = 1.30 <br>
d = 1/u 

In [3]:
n = 3  # Depth of binomial tree 
tdelta = tte / (n * 365)  # Time delta per one step (as fraction of year)
u = 1.30  
d = 1 / u  
rf = 0  
dy = 0  
pu = (1 + rf - dy - d) / (u - d)  
pd = 1 - pu  

assert side in [call, put] and style in [american, european]
print('%18s : %0.8f' % ('Node prob U', pu))
print('%18s : %0.8f' % ('Node prob D', pd))
print('%18s : %0.8f' % ('Node tdelta', tdelta))
print('%18s : %0.8f' % ('Node discount f', rf))

       Node prob U : 0.43478261
       Node prob D : 0.56521739
       Node tdelta : 0.33333333
   Node discount f : 0.00000000


In [4]:
print('Binomial Tree')

# Generate terminal nodes of binomial tree
level = []
print()
for j in range(0, n + 1):
    pr = price * d ** j * u ** (n - j)
    # Option value at the node (depending on side)
    ov = max(0.0, pr - strike) if side == call else max(0.0, strike - pr)
    level.append((pr, ov))
    print('Node [%i,%i] \t Price %6.3f \t Option Value %6.3f' % (n, j, pr, ov))

levels = [None, None, None]

# reduce binomial tree
for i in range(n - 1, -1, -1):  # [n-1 to 0]
    levelNext = []
    print()
    for j in range(0, i + 1):  
        node_u, node_d = level[j], level[j + 1]
        # Instrument's price at the node
        pr = node_d[0] / d
        # Option value at the node (depending on side)
        ov = (node_d[1] * pd + node_u[1] * pu) / (1 + rf)
        if style == american:  # for American options - not relevant for this assignment
            ov = max(ov, pr - strike if side == call else strike - pr)
        levelNext.append((pr, ov))
        print('Node [%i,%i] \t Price %6.3f \t Option Value %6.3f' % (i, j, pr, ov))
    level = levelNext
    if j <= 2: levels[j] = level 

Binomial Tree

Node [3,0] 	 Price 208.715 	 Option Value 103.715
Node [3,1] 	 Price 123.500 	 Option Value 18.500
Node [3,2] 	 Price 73.077 	 Option Value  0.000
Node [3,3] 	 Price 43.241 	 Option Value  0.000

Node [2,0] 	 Price 160.550 	 Option Value 55.550
Node [2,1] 	 Price 95.000 	 Option Value  8.043
Node [2,2] 	 Price 56.213 	 Option Value  0.000

Node [1,0] 	 Price 123.500 	 Option Value 28.698
Node [1,1] 	 Price 73.077 	 Option Value  3.497

Node [0,0] 	 Price 95.000 	 Option Value 14.454


Value of the derivative at each node can be seen in the last column above

Value of the call option at time 0 = 14.454

### Price a European Put Option
**Same parameters as the first part but now with N=2**

In [5]:
side = put  # Option side
style = european  # Option style
price = 95  #Current instrument price
strike = 105  #Strike price
riskfree = 0  
divyield = 0
tte = 365  # Time to expiration in days

print('Calculation Inputs')
print('%18s : %0.3f' % ('Price', price))
print('%18s : %0.3f' % ('Strike', strike))
print('%18s : %0.3f' % ('Risk-free', riskfree))
print('%18s : %0.3f' % ('Div Yield', divyield))
print('%18s : %0.3f' % ('TTE Days', tte))
print()

Calculation Inputs
             Price : 95.000
            Strike : 105.000
         Risk-free : 0.000
         Div Yield : 0.000
          TTE Days : 365.000



In [6]:
n = 2  # Depth of binomial tree (levels are numbered from 0 to n)
tdelta = tte / (n * 365)  # Time delta per one step (as fraction of year)
u = 1.30  # Up movement per step
d = 1 / u  # Down movement per step
rf = 0  # Risk-free rate per step
dy = 0  # Dividend yield per step
pu = (1 + rf - dy - d) / (u - d)  # Probability of up movement
pd = 1 - pu  # Probability of down movement

assert side in [call, put] and style in [american, european]
print('%18s : %0.8f' % ('Node prob U', pu))
print('%18s : %0.8f' % ('Node prob D', pd))
print('%18s : %0.8f' % ('Node tdelta', tdelta))
print('%18s : %0.8f' % ('Node discount f', rf))

       Node prob U : 0.43478261
       Node prob D : 0.56521739
       Node tdelta : 0.50000000
   Node discount f : 0.00000000


In [7]:
print('Binomial Tree')

# Generate terminal nodes of binomial tree
level = []
print()
for j in range(0, n + 1):
    pr = price * d ** j * u ** (n - j)
    # Option value at the node (depending on side)
    ov = max(0.0, pr - strike) if side == call else max(0.0, strike - pr)
    level.append((pr, ov))
    print('Node [%i,%i] \t Price %6.3f \t Option Value %6.3f' % (n, j, pr, ov))

levels = [None, None, None]

# reduce binomial tree
for i in range(n - 1, -1, -1):  # [n-1 to 0]
    levelNext = []
    print()
    for j in range(0, i + 1):  
        node_u, node_d = level[j], level[j + 1]
        # Instrument's price at the node
        pr = node_d[0] / d
        # Option value at the node (depending on side)
        ov = (node_d[1] * pd + node_u[1] * pu) / (1 + rf)
        if style == american:  # for American options - not relevant for this assignment
            ov = max(ov, pr - strike if side == call else strike - pr)
        levelNext.append((pr, ov))
        print('Node [%i,%i] \t Price %6.3f \t Option Value %6.3f' % (i, j, pr, ov))
    level = levelNext
    if j <= 2: levels[j] = level 

Binomial Tree

Node [2,0] 	 Price 160.550 	 Option Value  0.000
Node [2,1] 	 Price 95.000 	 Option Value 10.000
Node [2,2] 	 Price 56.213 	 Option Value 48.787

Node [1,0] 	 Price 123.500 	 Option Value  5.652
Node [1,1] 	 Price 73.077 	 Option Value 31.923

Node [0,0] 	 Price 95.000 	 Option Value 20.501


The price of the put option at time = 0 is 20.50

### Replicating Portfolio
**Constructing a Table that includes, for each price path and each t when it corresponds, the information on stock price evolution** 
* 𝑋𝑡(𝜔)
* Value of the option
* 𝑉𝑡𝐻(𝜔)
* Payoff of the option
* 𝐻(𝜔)
* Hedging strategy, 𝜑𝑡𝐻

In [8]:
import pandas as pd 

In [10]:
df = pd.read_excel(r'Replicating1.xlsx')
df = df.set_index('w')
df=df.rename(columns = {'Unnamed: 5':' '})
df = df.rename(columns = {'X0(w)':'X$_{0}$ (𝜔)'})
df = df.rename(columns = {'X1(w)':'X$_{1}$ (𝜔)'})
df = df.rename(columns = {'X2(w)':'X$_{2}$ (𝜔)'})

df = df.rename(columns = {'V-0H (w)':'V$_{0}$$^{H}$ (𝜔)'})
df = df.rename(columns = {'V-1H (w)':'V$_{1}$$^{H}$ (𝜔)'})
df = df.rename(columns = {'V-2H (w)':'V$_{2}$$^{H}$ (𝜔)'})

df

X$_{0}$ (𝜔)  X$_{1}$ (𝜔)  X$_{2}$ (𝜔)   H(w)      V$_{0}$$^{H}$ (𝜔)  \
w                                                                            
u,u             95       123.50       160.55   0.00 NaN               20.5   
u, d            95       123.50        95.00  10.00 NaN               20.5   
d, u            95        73.07        95.00  10.00 NaN               20.5   
d, d            95        73.07        56.21  48.78 NaN               20.5   

       V$_{1}$$^{H}$ (𝜔)  V$_{2}$$^{H}$ (𝜔)  
w                                            
u,u                31.92               0.00  
u, d               31.92              10.00  
d, u                5.65              10.00  
d, d                5.65              48.79

Now calculating the replicating strategy 

In [11]:
df1 = pd.read_excel(r'Replicating2.xlsx')
df1 = df1.set_index('w')
df1 = df1.rename(columns = {'k - 1H (w)':'𝜑$^{1}$$_{H}$ (𝜔)'})
df1 = df1.rename(columns = {'k - 2H (w)':'𝜑$^{2}$$_{H}$ (𝜔)'})
df1 

𝜑$^{1}$$_{H}$ (𝜔)  𝜑$^{2}$$_{H}$ (𝜔)
w                                          
u,u             0.400702          -0.861538
u, d            0.400702           0.769123
d, u            0.677155           0.198358
d, d            0.677155          -2.558719